src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,733 @@
+'''
+Created on 18 Mar 2020
+
+@author: Philip Winkler
+
+Copyright © 2020, Philip Winkler, Delmic
+
+This file is part of Odemis.
+
+Odemis is free software: you can redistribute it and/or modify it under the terms
+of the GNU General Public License version 2 as published by the Free Software
+Foundation.
+
+Odemis is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+Odemis. If not, see http://www.gnu.org/licenses/.
+'''
+from __future__ import division
+
+import glob
+import logging
+from odemis import model
+from odemis.model import Actuator, CancellableThreadPoolExecutor
+from odemis.util import to_str_escape
+import os
+import threading
+import time
+import serial
+from threading import Thread
+import copy
+from odemis.model import HwError
+from odemis.util.driver import getSerialDriver
+
+BAUDRATE = 115200
+# TODO: Is this always the same for all encoders? Otherwise, it needs to be a parameters in the yaml file.
+# WFM_STEPSIZE = 5e-6  # m / step
+USTEPS_PER_WFM = 8192
+# TODO: set encoder stepsize to proper value
+ENCODER_STEPSIZE = 1e-9  # m / step
+DEFAULT_AXIS_SPEED = 0.01  # m / s
+
+EOL = b'\r'  # 0x0D, carriage return
+sEOL = '\r'  # string version of EOL
+
+# Waveforms
+WAVEFORM_RHOMB = 1  # fast max speed
+WAVEFORM_DELTA = 2  # preferred, higher accuracy
+WAVEFORM_PARK = 4  # power off
+
+class PMDError(Exception):
+    def __init__(self, errno, strerror, *args, **kwargs):
+        super(PMDError, self).__init__(errno, strerror, *args, **kwargs)
+        self.args = (errno, strerror)
+        self.errno = errno
+        self.strerror = strerror
+
+    def __str__(self):
+        return self.args[1]
+
+
+class PMD401Bus(Actuator):
+    """
+    This represents the PMD401 motor controller bus for the PiezoMotor LEGS motors. The
+    three motor controllers (x, y, and z axes) are connected in a daisy chain. Only the first
+    controller is directly connected to the computer.
+    The specification for the hardware interface can be found in the document
+     "PiezoMotor_PMD401_Technical_Manual.pdf".
+
+    The class provides only part of the functionality of the model.Actuator interface. It fully
+    implements the functions moveRel and stop. However, there is no reference function and
+    moveAbs is only possible if we are operating in closed loop, i.e. an encoder (stage position sensor)
+    is present.
+    """
+
+    def __init__(self, name, role, port, axes, inverted=None, **kwargs):
+        """
+        :param axes (dict: {"x", "y", "z"} --> dict): axis name --> axis parameters (axis_number, range, unit, closed_loop, speed, wfm_stepsize)
+            axis_number (0 <= int <= 127), required (typically 1-3 for x-z)
+            wfm stepsize (int), stepsize (needs to be
```

Comment:
That docstring needs some love ;-) I don't think it's an int. Also indicate what "wfm" stands for ("waveform").